File: packages/cursor/scripts/lib/scan-sync-inbox.py
#!/usr/bin/env python3
"""Scan projetos do registry hostdime por trabalho não commitado + contexto recente."""
from __future__ import annotations

import json
import os
import re
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path


def project_slug(path: str) -> str:
    return re.sub(r"[^a-zA-Z0-9]+", "-", path.strip("/")).strip("-").lower()


def find_transcript_dir(cursor_dir: Path, repo_path: str) -> Path | None:
    projects = cursor_dir / "projects"
    if not projects.is_dir():
        return None
    slug = project_slug(repo_path)
    candidates = sorted(projects.glob(f"*{slug}*"), key=lambda p: p.stat().st_mtime, reverse=True)
    for cand in candidates:
        transcripts = cand / "agent-transcripts"
        if transcripts.is_dir():
            return transcripts
    return None


def latest_user_query(transcript_dir: Path) -> str | None:
    files = sorted(
        transcript_dir.rglob("*.jsonl"),
        key=lambda p: p.stat().st_mtime,
        reverse=True,
    )
    for fp in files[:5]:
        try:
            for line in fp.read_text(encoding="utf-8", errors="replace").splitlines():
                if not line.strip():
                    continue
                row = json.loads(line)
                if row.get("role") != "user":
                    continue
                msg = row.get("message") or {}
                parts = msg.get("content") or []
                for part in parts:
                    if part.get("type") != "text":
                        continue
                    text = part.get("text") or ""
                    text = re.sub(r"<user_query>\s*", "", text)
                    text = re.sub(r"\s*</user_query>", "", text)
                    text = text.strip()
                    if text and len(text) > 10:
                        return text[:280]
        except (OSError, json.JSONDecodeError):
            continue
    return None


def git_info(repo: Path) -> dict | None:
    if not (repo / ".git").exists():
        return None
    try:
        subprocess.run(
            ["git", "-C", str(repo), "rev-parse", "--git-dir"],
            capture_output=True,
            check=True,
        )
    except (OSError, subprocess.CalledProcessError):
        return None

    def run(args: list[str]) -> str:
        proc = subprocess.run(
            ["git", "-C", str(repo), *args],
            capture_output=True,
            text=True,
            check=False,
        )
        return (proc.stdout or "").strip()

    porcelain = run(["status", "--porcelain"])
    if not porcelain:
        return None

    branch = run(["branch", "--show-current"]) or "(detached)"
    last_commit = run(["log", "-1", "--pretty=%s"])
    changed = []
    for line in porcelain.splitlines():
        if len(line) < 4:
            continue
        path = line[3:].strip()
        if " -> " in path:
            path = path.split(" -> ", 1)[1]
        changed.append({"code": line[:2], "path": path})

    staged = sum(1 for c in changed if c["code"][0] != " " and c["code"][0] != "?")
    unstaged = sum(1 for c in changed if c["code"][1] != " ")

    return {
        "branch": branch,
        "lastCommit": last_commit,
        "changedCount": len(changed),
        "stagedCount": staged,
        "unstagedCount": unstaged,
        "changedFiles": [c["path"] for c in changed[:12]],
        "porcelain": porcelain,
    }


def infer_objective(git: dict, last_query: str | None) -> str:
    parts: list[str] = []
    if last_query:
        parts.append(f"Chat recente: {last_query}")
    if git.get("lastCommit"):
        parts.append(f"Último commit: {git['lastCommit']}")
    files = git.get("changedFiles") or []
    if files:
        parts.append(f"Arquivos: {', '.join(files[:5])}")
    return " · ".join(parts) if parts else "Alterações locais não commitadas"


def scan_projects(registry_path: Path, cursor_dir: Path) -> list[dict]:
    if not registry_path.is_file():
        return []
    data = json.loads(registry_path.read_text(encoding="utf-8"))
    results: list[dict] = []
    seen: set[str] = set()

    for entry in data.get("projects", []):
        path = entry.get("path", "")
        if not path:
            continue
        repo = Path(path).resolve()
        key = str(repo)
        if key in seen or not repo.is_dir():
            continue
        seen.add(key)

        git = git_info(repo)
        if not git:
            continue

        transcripts = find_transcript_dir(cursor_dir, str(repo))
        last_query = latest_user_query(transcripts) if transcripts else None

        results.append(
            {
                "path": str(repo),
                "name": repo.name,
                "branch": git["branch"],
                "changedCount": git["changedCount"],
                "stagedCount": git["stagedCount"],
                "unstagedCount": git["unstagedCount"],
                "changedFiles": git["changedFiles"],
                "lastCommit": git["lastCommit"],
                "lastUserQuery": last_query,
                "objective": infer_objective(git, last_query),
                "scannedAt": datetime.now(timezone.utc).isoformat(),
            }
        )

    results.sort(key=lambda r: r["changedCount"], reverse=True)
    return results


def main() -> int:
    cursor_dir = Path(os.environ.get("CURSOR_USER_DIR", Path.home() / ".cursor"))
    registry = cursor_dir / "hostdime-ia" / "projects.json"
    out_path = cursor_dir / "hostdime-ia" / "sync-inbox.json"

    if len(sys.argv) > 1 and sys.argv[1] == "--registry":
        registry = Path(sys.argv[2])
    if len(sys.argv) > 1 and sys.argv[1] == "--out":
        out_path = Path(sys.argv[2])

    items = scan_projects(registry, cursor_dir)
    payload = {"version": 1, "generatedAt": datetime.now(timezone.utc).isoformat(), "items": items}
    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text(json.dumps(payload, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    print(json.dumps(payload, ensure_ascii=False))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: packages/cursor/scripts/lib/sync-inbox/ts/scan-sync-inbox.ts
#!/usr/bin/env node
/** Scan projetos do registry hostdime por trabalho não commitado + contexto recente. */
import { existsSync, mkdirSync, readFileSync, readdirSync, statSync, writeFileSync } from 'node:fs';
import { basename, join, resolve, sep } from 'node:path';
import { homedir } from 'node:os';
import { spawnSync } from 'node:child_process';
import { pathToFileURL } from 'node:url';

const TOOLING_MARKERS = ['.gitignore', '.cursor/', 'package-lock.json', 'pnpm-lock.yaml', 'yarn.lock'];

const CONVENTIONAL_PREFIX = /^(feat|fix|chore|refactor|docs|test|style|perf|ci|build)(\([^)]+\))?:\s*/i;

interface GitInfo {
  branch: string;
  lastCommit: string;
  changedCount: number;
  stagedCount: number;
  unstagedCount: number;
  changedFiles: string[];
  porcelain: string;
}

interface ScanItem {
  path: string;
  name: string;
  branch: string;
  changedCount: number;
  stagedCount: number;
  unstagedCount: number;
  changedFiles: string[];
  recentFiles: string[];
  lastCommit: string;
  lastUserQuery: string | null;
  summary: string;
  detail: string;
  objective: string;
  scannedAt: string;
}

function cursorProjectDirName(repoPath: string): string {
  return resolve(repoPath).replaceAll(sep, '-').replace(/^-/, '');
}

function projectSlug(path: string): string {
  return path
    .trim()
    .replace(/^\/+|\/+$/g, '')
    .replace(/[^a-zA-Z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .toLowerCase();
}

function findTranscriptDir(cursorDir: string, repoPath: string): string | null {
  const projects = join(cursorDir, 'projects');
  if (!existsSync(projects)) return null;

  const exact = join(projects, cursorProjectDirName(repoPath), 'agent-transcripts');
  if (existsSync(exact)) return exact;

  const slug = projectSlug(repoPath);
  const repoName = basename(repoPath).toLowerCase();
  const candidates: string[] = [];

  for (const entry of readdirSync(projects, { withFileTypes: true })) {
    if (!entry.isDirectory()) continue;
    const name = entry.name.toLowerCase();
    if (name.includes(slug) || name.endsWith(`-${repoName}`)) {
      const cand = join(projects, entry.name, 'agent-transcripts');
      if (existsSync(cand)) candidates.push(cand);
    }
  }

  if (candidates.length === 0) return null;
  candidates.sort((a, b) => statSync(b).mtimeMs - statSync(a).mtimeMs);
  return candidates[0]!;
}

function loadRecentFiles(cursorDir: string, repoPath: string, limit = 5): string[] {
  const stateFile = join(cursorDir, 'ide_state.json');
  if (!existsSync(stateFile)) return [];

  let data: { recentlyViewedFiles?: Array<{ absolutePath?: string }> };
  try {
    data = JSON.parse(readFileSync(stateFile, 'utf-8')) as typeof data;
  } catch {
    return [];
  }

  const repo = resolve(repoPath);
  const prefix = `${repo}${sep}`;
  const files: string[] = [];

  for (const entry of data.recentlyViewedFiles ?? []) {
    const absolute = entry.absolutePath ?? '';
    if (!absolute.startsWith(prefix)) continue;
    const rel = absolute.slice(prefix.length);
    if (rel && !files.includes(rel)) files.push(rel);
    if (files.length >= limit) break;
  }
  return files;
}

function isNoiseQuery(text: string): boolean {
  const lower = text.toLowerCase();
  const noiseMarkers = [
    'responda somente em json',
    'analise o contexto de retomada',
    '{"doing":',
    '<user_info>',
    'you are auto, an agent router',
  ];
  return noiseMarkers.some((marker) => lower.includes(marker));
}

function cleanUserQuery(text: string): string {
  let cleaned = text
    .replace(/<user_query>\s*/g, '')
    .replace(/\s*<\/user_query>/g, '')
    .replace(/<image_files>.*?<\/image_files>/gs, '')
    .replace(/\[Image\]\s*/g, '')
    .replace(/\s+/g, ' ')
    .trim();
  if (!cleaned) return '';
  const first = cleaned.split(/(?<=[.!?])\s+|\n+/)[0]!.trim();
  return first.length > 120 ? `${first.slice(0, 119)}…` : first;
}

function walkJsonlFiles(dir: string): string[] {
  const results: string[] = [];
  const stack = [dir];
  while (stack.length > 0) {
    const current = stack.pop()!;
    for (const entry of readdirSync(current, { withFileTypes: true })) {
      const full = join(current, entry.name);
      if (entry.isDirectory()) stack.push(full);
      else if (entry.name.endsWith('.jsonl')) results.push(full);
    }
  }
  return results;
}

function latestUserQuery(transcriptDir: string): string | null {
  const files = walkJsonlFiles(transcriptDir)
    .sort((a, b) => statSync(b).mtimeMs - statSync(a).mtimeMs)
    .slice(0, 8);

  for (const fp of files) {
    try {
      const lines = readFileSync(fp, 'utf-8').split('\n').filter(Boolean).reverse();
      for (const line of lines) {
        const row = JSON.parse(line) as {
          role?: string;
          message?: { content?: Array<{ type?: string; text?: string }> };
        };
        if (row.role !== 'user') continue;
        for (const part of row.message?.content ?? []) {
          if (part.type !== 'text') continue;
          const text = cleanUserQuery(part.text ?? '');
          if (text.length > 10 && !isNoiseQuery(text)) return text;
        }
      }
    } catch {
      continue;
    }
  }
  return null;
}

function shortenCommit(message: string, maxLen = 72): string {
  const text = message.trim().replace(CONVENTIONAL_PREFIX, '');
  if (text.length <= maxLen) return text;
  return `${text.slice(0, maxLen - 1).trim()}…`;
}

function basenameOnly(path: string): string {
  if (path.endsWith('/')) return `${basename(path.replace(/\/$/, ''))}/`;
  return basename(path);
}

function inferAreas(changedFiles: string[]): string[] {
  const areas: string[] = [];
  for (const raw of changedFiles) {
    const path = raw.replace(/\\/g, '/');
    const module = path.match(/\/modules\/([^/]+)\//);
    if (module) {
      areas.push(module[1]!);
      continue;
    }
    if (path.includes('/components/')) {
      const part = basename(path).replace(/\.[^.]+$/, '');
      if (!['index', 'types'].includes(part)) areas.push(part);
      continue;
    }
    if (path.includes('/doc/') || path.endsWith('.md')) {
      areas.push('documentação');
      continue;
    }
    if (path.toLowerCase().includes('test')) {
      areas.push('testes');
      continue;
    }
    const name = basenameOnly(path);
    if (!areas.includes(name)) areas.push(name);
  }

  const deduped: string[] = [];
  const seen = new Set<string>();
  for (const area of areas) {
    const key = area.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    deduped.push(area);
  }
  return deduped.slice(0, 4);
}

function isToolingOnly(changedFiles: string[]): boolean {
  if (changedFiles.length === 0) return false;
  for (const raw of changedFiles) {
    const path = raw.replace(/\\/g, '/');
    if (TOOLING_MARKERS.some((marker) => path.includes(marker))) continue;
    if (path.startsWith('.cursor/')) continue;
    return false;
  }
  return true;
}

function featureBranchLabel(branch: string): string | null {
  if (!branch || ['main', 'master', 'develop', 'dev', '(detached)'].includes(branch)) {
    return null;
  }
  return branch.replace(/-/g, ' ').replace(/_/g, ' ');
}

function buildResume(params: {
  branch: string;
  changedCount: number;
  changedFiles: string[];
  lastCommit: string | null;
  lastQuery: string | null;
  recentFiles: string[];
}): [string, string, string] {
  const { branch, changedCount, changedFiles, lastCommit, lastQuery, recentFiles } = params;
  const areas = inferAreas(changedFiles.length > 0 ? changedFiles : recentFiles);
  const branchHint = featureBranchLabel(branch);
  const n = changedCount;

  let summary: string;
  if (lastQuery) {
    summary = `Pedido: ${lastQuery}`;
  } else if (isToolingOnly(changedFiles)) {
    summary = 'Setup Cursor / sync do projeto';
  } else if (areas.length > 0 && !['documentação', 'testes', '.gitignore', 'gitignore'].includes(areas[0]!)) {
    const lead = areas[0]!;
    summary =
      branchHint && !lead.toLowerCase().includes(branchHint.toLowerCase())
        ? `${branchHint} — ${lead}`
        : lead;
    if (areas.length > 1) summary += ` (+${areas.length - 1})`;
  } else if (branchHint) {
    summary = `Branch ${branchHint}`;
  } else if (lastCommit) {
    summary = shortenCommit(lastCommit);
  } else {
    summary = `${n} arquivo(s) em andamento`;
  }

  const detailParts: string[] = [];
  if (branch) detailParts.push(branch);
  detailParts.push(`${n} arquivo${n !== 1 ? 's' : ''}`);
  if (areas.length > 0) detailParts.push(areas.slice(0, 2).join(', '));
  else if (recentFiles.length > 0) detailParts.push(basenameOnly(recentFiles[0]!));
  const detail = detailParts.join(' · ');

  const objectiveParts: string[] = [];
  if (lastQuery) objectiveParts.push(`Último pedido: ${lastQuery}`);
  if (branchHint) objectiveParts.push(`Branch: ${branch}`);
  if (areas.length > 0) objectiveParts.push(`Áreas: ${areas.slice(0, 3).join(', ')}`);
  else if (changedFiles.length > 0) {
    objectiveParts.push(`Arquivos: ${changedFiles.slice(0, 4).map(basenameOnly).join(', ')}`);
  }
  if (lastCommit && !lastQuery) objectiveParts.push(`Commit: ${shortenCommit(lastCommit, 60)}`);
  const objective = objectiveParts.length > 0 ? objectiveParts.join(' · ') : summary;

  return [summary, detail, objective];
}

function gitInfo(repo: string): GitInfo | null {
  if (!existsSync(join(repo, '.git'))) return null;

  const run = (args: string[]): string => {
    const proc = spawnSync('git', ['-C', repo, ...args], { encoding: 'utf-8' });
    return (proc.stdout ?? '').trim();
  };

  try {
    const check = spawnSync('git', ['-C', repo, 'rev-parse', '--git-dir'], { encoding: 'utf-8' });
    if (check.status !== 0) return null;
  } catch {
    return null;
  }

  const porcelain = run(['status', '--porcelain']);
  if (!porcelain) return null;

  const branch = run(['branch', '--show-current']) || '(detached)';
  const lastCommit = run(['log', '-1', '--pretty=%s']);

  const changed: Array<{ code: string; path: string }> = [];
  for (const line of porcelain.split('\n')) {
    if (line.length < 4) continue;
    let path = line.slice(3).trim();
    if (path.includes(' -> ')) path = path.split(' -> ')[1] ?? path;
    changed.push({ code: line.slice(0, 2), path });
  }

  const staged = changed.filter((c) => c.code[0] !== ' ' && c.code[0] !== '?').length;
  const unstaged = changed.filter((c) => c.code[1] !== ' ').length;

  return {
    branch,
    lastCommit,
    changedCount: changed.length,
    stagedCount: staged,
    unstagedCount: unstaged,
    changedFiles: changed.slice(0, 12).map((c) => c.path),
    porcelain,
  };
}

export function scanProjects(registryPath: string, cursorDir: string): ScanItem[] {
  if (!existsSync(registryPath)) return [];

  const data = JSON.parse(readFileSync(registryPath, 'utf-8')) as {
    projects?: Array<{ path?: string }>;
  };
  const results: ScanItem[] = [];
  const seen = new Set<string>();

  for (const entry of data.projects ?? []) {
    const path = entry.path ?? '';
    if (!path) continue;
    const repo = resolve(path);
    const key = repo;
    if (seen.has(key) || !existsSync(repo)) continue;
    seen.add(key);

    const git = gitInfo(repo);
    if (!git) continue;

    const transcripts = findTranscriptDir(cursorDir, repo);
    const lastQuery = transcripts ? latestUserQuery(transcripts) : null;
    const recentFiles = loadRecentFiles(cursorDir, repo);
    const [summary, detail, objective] = buildResume({
      branch: git.branch,
      changedCount: git.changedCount,
      changedFiles: git.changedFiles,
      lastCommit: git.lastCommit,
      lastQuery,
      recentFiles,
    });

    results.push({
      path: repo,
      name: basename(repo),
      branch: git.branch,
      changedCount: git.changedCount,
      stagedCount: git.stagedCount,
      unstagedCount: git.unstagedCount,
      changedFiles: git.changedFiles,
      recentFiles,
      lastCommit: git.lastCommit,
      lastUserQuery: lastQuery,
      summary,
      detail,
      objective,
      scannedAt: new Date().toISOString(),
    });
  }

  results.sort((a, b) => {
    const hasChatA = a.lastUserQuery ? 0 : 1;
    const hasChatB = b.lastUserQuery ? 0 : 1;
    const toolingA = isToolingOnly(a.changedFiles) ? 0 : 1;
    const toolingB = isToolingOnly(b.changedFiles) ? 0 : 1;
    if (hasChatA !== hasChatB) return hasChatA - hasChatB;
    if (toolingA !== toolingB) return toolingA - toolingB;
    return b.changedCount - a.changedCount;
  });

  return results;
}

function main(): number {
  const cursorDir = process.env.CURSOR_USER_DIR ?? join(homedir(), '.cursor');
  let registry = join(cursorDir, 'hostdime-ia/projects.json');
  let outPath = join(cursorDir, 'hostdime-ia/sync-inbox.json');

  if (process.argv.length > 2 && process.argv[2] === '--registry') {
    registry = process.argv[3]!;
  }
  if (process.argv.length > 2 && process.argv[2] === '--out') {
    outPath = process.argv[3]!;
  }

  const items = scanProjects(registry, cursorDir);
  const payload = {
    version: 1,
    generatedAt: new Date().toISOString(),
    items,
  };

  mkdirSync(join(outPath, '..'), { recursive: true });
  writeFileSync(outPath, `${JSON.stringify(payload, null, 2)}\n`, 'utf-8');
  process.stdout.write(JSON.stringify(payload));
  return 0;
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? '').href) {
  process.exit(main());
}
